ssl/record/methods/tls_common.c

```diff
@@ -0,0 +1,1408 @@
+                return OSSL_RECORD_RETURN_FATAL;
+            }
+            sslv2pkt = pkt;
+            if (!PACKET_get_net_2_len(&sslv2pkt, &sslv2len)
+                    || !PACKET_get_1(&sslv2pkt, &type)) {
+                RLAYERfatal(rl, SSL_AD_DECODE_ERROR, ERR_R_INTERNAL_ERROR);
+                return OSSL_RECORD_RETURN_FATAL;
+            }
+            /*
+             * The first record received by the server may be a V2ClientHello.
+             */
+            if (rl->role == OSSL_RECORD_ROLE_SERVER
+                    && rl->is_first_record
+                    && (sslv2len & 0x8000) != 0
+                    && (type == SSL2_MT_CLIENT_HELLO)) {
+                /*
+                 *  SSLv2 style record
+                 *
+                 * |num_recs| here will actually always be 0 because
+                 * |num_recs > 0| only ever occurs when we are processing
+                 * multiple app data records - which we know isn't the case here
+                 * because it is an SSLv2ClientHello. We keep it using
+                 * |num_recs| for the sake of consistency
+                 */
+                thisrr->type = SSL3_RT_HANDSHAKE;
+                thisrr->rec_version = SSL2_VERSION;
+
+                thisrr->length = sslv2len & 0x7fff;
+
+                if (thisrr->length > SSL3_BUFFER_get_len(rbuf)
+                                     - SSL2_RT_HEADER_LENGTH) {
+                    RLAYERfatal(rl, SSL_AD_RECORD_OVERFLOW,
+                                SSL_R_PACKET_LENGTH_TOO_LONG);
+                    return OSSL_RECORD_RETURN_FATAL;
+                }
+            } else {
+                /* SSLv3+ style record */
+
+                /* Pull apart the header into the SSL3_RECORD */
+                if (!PACKET_get_1(&pkt, &type)
+                        || !PACKET_get_net_2(&pkt, &version)
+                        || !PACKET_get_net_2_len(&pkt, &thisrr->length)) {
+                    rl->msg_callback(0, 0, SSL3_RT_HEADER, p, 5, rl->cbarg);
+                    RLAYERfatal(rl, SSL_AD_DECODE_ERROR, ERR_R_INTERNAL_ERROR);
+                    return OSSL_RECORD_RETURN_FATAL;
+                }
+                thisrr->type = type;
+                thisrr->rec_version = version;
+
+                /*
+                 * When we call validate_record_header() only records actually
+                 * received in SSLv2 format should have the record version set
+                 * to SSL2_VERSION. This way validate_record_header() can know
+                 * what format the record was in based on the version.
+                 */
+                if (thisrr->rec_version == SSL2_VERSION) {
+                    RLAYERfatal(rl, SSL_AD_PROTOCOL_VERSION,
+                                SSL_R_WRONG_VERSION_NUMBER);
+                    return OSSL_RECORD_RETURN_FATAL;
+                }
+
+                rl->msg_callback(0, version, SSL3_RT_HEADER, p, 5, rl->cbarg);
```

Comment:
I guess this is always non-NULL?

Reply:
It always is (libssl will always ensure this) - but strictly speaking the record layer should check this itself (which it doesn't). I've pushed a commit to do that.